File: krishideep_app/lib/models/crop_recommendation.dart
import 'crop.dart';

class CropRecommendation {
  final List<Crop> recommendedCrops;
  final String recommendation;
  final String confidenceLevel;
  final DateTime timestamp;
  final Map<String, dynamic> analysisData;

  CropRecommendation({
    required this.recommendedCrops,
    required this.recommendation,
    required this.confidenceLevel,
    required this.timestamp,
    required this.analysisData,
  });

  Map<String, dynamic> toJson() {
    return {
      'recommendedCrops':
          recommendedCrops.map((crop) => crop.toJson()).toList(),
      'recommendation': recommendation,
      'confidenceLevel': confidenceLevel,
      'timestamp': timestamp.toIso8601String(),
      'analysisData': analysisData,
    };
  }

  factory CropRecommendation.fromJson(Map<String, dynamic> json) {
    return CropRecommendation(
      recommendedCrops: (json['recommendedCrops'] as List)
          .map((crop) => Crop.fromJson(crop))
          .toList(),
      recommendation: json['recommendation'],
      confidenceLevel: json['confidenceLevel'],
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'])
          : DateTime.now(),
      analysisData: json['analysisData'],
    );
  }
}

File: krisideepweb/src/models/CropRecommendation.ts
import { Crop } from './Crop';

export interface CropRecommendation {
  recommendedCrops: Crop[];
  recommendation: string;
  confidenceLevel: string;
  timestamp: Date;
  analysisData: Record<string, any>;
}

export interface CropRecommendationJson {
  recommendedCrops: any[];
  recommendation: string;
  confidenceLevel: string;
  timestamp: string;
  analysisData: Record<string, any>;
}

export const cropRecommendationFromJson = (json: CropRecommendationJson): CropRecommendation => ({
  recommendedCrops: json.recommendedCrops.map((crop: any) => ({
    id: crop.id,
    name: crop.name,
    nameHindi: crop.nameHindi,
    description: crop.description,
    expectedYield: crop.expectedYield,
    profitMargin: crop.profitMargin,
    sustainabilityScore: crop.sustainabilityScore,
    suitableConditions: crop.suitableConditions,
    imageUrl: crop.imageUrl || '',
  })),
  recommendation: json.recommendation,
  confidenceLevel: json.confidenceLevel,
  timestamp: new Date(json.timestamp),
  analysisData: json.analysisData,
});

export const cropRecommendationToJson = (recommendation: CropRecommendation): CropRecommendationJson => ({
  recommendedCrops: recommendation.recommendedCrops.map(crop => ({
    id: crop.id,
    name: crop.name,
    nameHindi: crop.nameHindi,
    description: crop.description,
    expectedYield: crop.expectedYield,
    profitMargin: crop.profitMargin,
    sustainabilityScore: crop.sustainabilityScore,
    suitableConditions: crop.suitableConditions,
    imageUrl: crop.imageUrl,
  })),
  recommendation: recommendation.recommendation,
  confidenceLevel: recommendation.confidenceLevel,
  timestamp: recommendation.timestamp.toISOString(),
  analysisData: recommendation.analysisData,
});
